spec/ruby/library/set/sortedset/filter_spec.rb

```diff
@@ -0,0 +1,35 @@
+require File.expand_path('../../../../spec_helper', __FILE__)
+require 'set'
+
+describe "SortedSet#filter!" do
+  before :each do
+    @set = SortedSet["one", "two", "three"]
+  end
+
+  it "yields each Object in self in sorted order" do
+    res = []
+    @set.filter! { |x| res << x }
+    res.should == ["one", "two", "three"].sort
+  end
+
+  it "keeps every element from self for which the passed block returns true" do
+    @set.filter! { |x| x.size != 3 }
+    @set.to_a.should == ["three"]
+  end
+
+  it "returns self when self was modified" do
+    @set.filter! { false }.should equal(@set)
+  end
+
+  it "returns nil when self was not modified" do
+    @set.filter! { true }.should be_nil
+  end
+
+  it "returns an Enumerator when passed no block" do
+    enum = @set.filter!
+    enum.should be_an_instance_of(Enumerator)
+
+    enum.each { |x| x.size != 3 }
+    @set.to_a.should == ["three"]
+  end
```

Comment:
Could you use shared specs with `Set#keep_if` and `Set#select`?
Those do not use share specs currently but it should be trivial to add it.

Reply:
Sorry, I'm not sure I understand. Do you mean `Set#keep_if`, `Set#select!`, and `Set#filter!` should have identical specs?
`keep_if` and `select!` have different behavior for sets and hashes, so I chose to make `filter!` mirror `select!` instead of `keep_if` where those diverged.
```ruby
s = SortedSet[1,2,3]
=> #<SortedSet: {1, 2, 3}>
s.keep_if { true }
=> #<SortedSet: {1, 2, 3}>
s.select! { true }
=> nil
```

Reply:
Right, but except the return value they behave identically.
So most specs for these 3 methods could be shared.
Or at least the specs of `select!` and `filter!` can be shared.

Reply:
Is there a way to adapt the spec of `select!` for `filter!` other than copying and pasting it? I wasn't able to find anything looking through other spec files.

Reply:
Yes, see for example `core/enumerable/shared/find_all.rb` which you reused in `spec/ruby/core/enumerable/filter_spec.rb` with `it_behaves_like`

Reply:
Thank you. I think I got it working. 👍